flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManagerDriver.java

```diff
@@ -0,0 +1,301 @@
+import org.apache.flink.configuration.GlobalConfiguration;
+import org.apache.flink.configuration.TaskManagerOptions;
+import org.apache.flink.kubernetes.configuration.KubernetesConfigOptions;
+import org.apache.flink.kubernetes.configuration.KubernetesResourceManagerConfiguration;
+import org.apache.flink.kubernetes.kubeclient.FlinkKubeClient;
+import org.apache.flink.kubernetes.kubeclient.factory.KubernetesTaskManagerFactory;
+import org.apache.flink.kubernetes.kubeclient.parameters.KubernetesTaskManagerParameters;
+import org.apache.flink.kubernetes.kubeclient.resources.KubernetesPod;
+import org.apache.flink.kubernetes.kubeclient.resources.KubernetesWatch;
+import org.apache.flink.kubernetes.utils.KubernetesUtils;
+import org.apache.flink.runtime.clusterframework.ApplicationStatus;
+import org.apache.flink.runtime.clusterframework.BootstrapTools;
+import org.apache.flink.runtime.clusterframework.ContaineredTaskManagerParameters;
+import org.apache.flink.runtime.clusterframework.TaskExecutorProcessSpec;
+import org.apache.flink.runtime.clusterframework.types.ResourceID;
+import org.apache.flink.runtime.concurrent.FutureUtils;
+import org.apache.flink.runtime.externalresource.ExternalResourceUtils;
+import org.apache.flink.runtime.resourcemanager.active.AbstractResourceManagerDriver;
+import org.apache.flink.runtime.resourcemanager.active.ResourceManagerDriver;
+import org.apache.flink.runtime.resourcemanager.exceptions.ResourceManagerException;
+import org.apache.flink.util.ExceptionUtils;
+import org.apache.flink.util.FlinkException;
+import org.apache.flink.util.Preconditions;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Implementation of {@link ResourceManagerDriver} for Kubernetes deployment.
+ */
+public class KubernetesResourceManagerDriver extends AbstractResourceManagerDriver<KubernetesWorkerNode>
+	implements FlinkKubeClient.PodCallbackHandler {
+
+	/** The taskmanager pod name pattern is {clusterId}-{taskmanager}-{attemptId}-{podIndex}. */
+	private static final String TASK_MANAGER_POD_FORMAT = "%s-taskmanager-%d-%d";
+
+	private final String clusterId;
+
+	private final FlinkKubeClient kubeClient;
+
+	/** Request resource futures, keyed by pod names. */
+	private final Map<String, CompletableFuture<KubernetesWorkerNode>> requestResourceFutures;
+
+	/** When ResourceManager failover, the max attempt should recover. */
+	private long currentMaxAttemptId = 0;
+
+	/** Current max pod index. When creating a new pod, it should increase one. */
+	private long currentMaxPodId = 0;
+
+	private KubernetesWatch podsWatch;
+
+	public KubernetesResourceManagerDriver(
+			Configuration flinkConfig,
+			FlinkKubeClient kubeClient,
+			KubernetesResourceManagerConfiguration configuration) {
+		super(flinkConfig, GlobalConfiguration.loadConfiguration());
```

Comment:
> The problem is that it is not guaranteed at this point that you can read a proper Flink configuration.
That sounds to be a problem. Could you explain what might prevent reading a proper Flink configuration?

> Moreover, you are missing all dynamically configured properties which are processed in the entrypoints.
I think that is exactly what we want. Please see my other comment about `flinkClientConfig`.

Reply:
The method relies on the fact that some environment variable is set. Depending on the how the process is setup this might be set or not. In the entrypoints we actually also consider a potentially configured configuration directory which is passed in as an argument. Hence, we might read completely different configurations here.

Reply:
Moreover, relying on env variables to read the right things will make the testing of this component much harder if we want to control the "base" configuration.